specifyweb/frontend/js_src/lib/components/AttachmentsBulkImport/utils.ts

```diff
@@ -0,0 +1,416 @@
+  uploadableFiles: RA<PartialUploadableFileSpec>,
+  uploadSpec: AttachmentUploadSpec,
+  keepDisambiguation: boolean = false
+): Promise<RA<PartialUploadableFileSpec>> {
+  const { baseTable, path } =
+    staticAttachmentImportPaths[uploadSpec.staticPathKey];
+
+  const validationQueryResource = generateInQueryResource(
+    baseTable,
+    path,
+    uploadableFiles.map(({ file }) => file?.parsedName),
+    'Batch Attachment Upload'
+  );
+  const validationResponse = filterArray(
+    (await validationPromiseGenerator(validationQueryResource)).map(
+      ({ targetId, restResult }) => {
+        const rawResult = uploadSpec.formatQueryResults(restResult[0]);
+        return rawResult === undefined ? undefined : { targetId, rawResult };
+      }
+    )
+  );
+
+  return matchFileSpec(uploadableFiles, validationResponse, keepDisambiguation);
+}
+
+export function matchSelectedFiles(
+  previousUploadables: RA<PartialUploadableFileSpec>,
+  filesToResolve: RA<PartialUploadableFileSpec>
+): RA<PartialUploadableFileSpec> {
+  let resolvedFiles = previousUploadables;
+  filesToResolve.forEach((uploadable) => {
+    const matchedIndex = resolvedFiles.findIndex((previousUploadable) => {
+      if (
+        previousUploadable.attachmentId !== undefined ||
+        previousUploadable.file instanceof File
+      )
+        return false;
+
+      return (
+        previousUploadable.file !== undefined &&
+        previousUploadable.file.name === uploadable.file.name &&
+        previousUploadable.file.size === uploadable.file.size &&
+        previousUploadable.file.type === uploadable.file.type
+      );
+    });
+
+    resolvedFiles =
+      matchedIndex === -1
+        ? insertItem(resolvedFiles, resolvedFiles.length, uploadable)
+        : replaceItem(resolvedFiles, matchedIndex, {
+            ...resolvedFiles[matchedIndex],
+            file: uploadable.file,
+            /*
+             * Generating tokens again because the file could have been
+             * uploaded to the asset server but not yet recorded in Specify DB.
+             */
+            uploadTokenSpec: undefined,
+            // Take the new status in case of parse failure was reported.
+            status: uploadable.status,
+          });
+  });
+  return resolvedFiles;
+}
+
+export function resolveFileNames(
+  previousFile: UnBoundFile,
+  getFormatted: (
+    rawName: number | string | null | undefined
+  ) => string | undefined,
+  formatter?: UiFormatter
+): PartialUploadableFileSpec {
+  let nameToParse: string | undefined;
+  if (
+    formatter === undefined ||
+    formatter.fields.some((field) => field instanceof formatterTypeMapper.regex)
+  )
+    nameToParse = stripFileExtension(previousFile.name);
+  else {
+    const formattedLength = formatter.fields.reduce(
+      (length, field) => length + field.size,
+      0
+    );
+    nameToParse = previousFile.name.slice(0, formattedLength);
+  }
+  previousFile.parsedName = f.maybe(nameToParse, getFormatted);
```

Comment:
I added this because it simplifies serializing the object. JSON conversion will preserve this, without me having to add a key. Also File() object doesn't have a native parsedName attribute

Reply:
ah I didn't notice that it is not a regular object, but an object created by the browser - this solution seems good in that case

Reply:
(optional)

though I would suggest using the "File" object less, and certainly not mutating it (not just because creating such an object from scratch is harder thus forcing you to use mutation).

The browser may define a native "parsedName" property on the "File" object in the future - that may break your code

i.e see https://developer.chrome.com/blog/smooshgate/

you could have a type like this instead:

```ts
type SomeType = { readonly file: File; readonly parsedName: string; ... };
```

Reply:
fixed